contributing/specifications/functions.md

```diff
@@ -66,6 +66,18 @@ A Function MUST NOT return a Warning or Normal result every time it is called. A
 Function SHOULD return Warning or Normal results to indicate transitions - e.g.
 when entering a warning or normal state.
 
+## External Systems
+
+A Function MAY interact with external systems, defined as systems other than the
+Crossplane process that initiated the RunFunctionRequest.
+
+A Function SHOULD NOT mutate the state of external systems directly. A Function
+MAY mutate the state of external systems indirectly, by composing Crossplane
+managed resources (MRs).
+
+A Function MUST NOT communicate directly with the Kubernetes API Server, for
+example to read or mutate API resources.
```

Comment:
> To help clarify this would it be worthwhile extending this statement to explicitly cover how communication with the Kubernetes API may be achieved?

I considered adding something like this. I didn't because "you can achieve it this way instead" felt less like a specification concern and more like a documentation concern. I see your point though, maybe worth it if it helps avoid confusion.

Regarding the thread, it seems the main concern is that the "MUST NOT communicate directly with the Kubernetes API Server" wording prevents a function accessing the API server to load IRSA credentials or similar, right?

That is a use case we want to support. I'm not sure how we will yet though. Some early thoughts:

* I don't think a function reading a ProviderConfig is the right path. A ProviderConfig should configure a provider (family), not functions.
* Whatever we do should work with https://github.com/crossplane/crossplane/pull/5543, e.g. I imagine a new function credentials source of type `IRSA`.
* Figuring out how this could work for functions that run outside a Kubernetes cluster will be tricky.

I understand the proposed update makes it impossible for a compliant function to use IRSA. My preference is to live with that for now. I expect we'll add language to relax this constraint in future once we know enough to be specific about what use cases (e.g. IRSA) we must allow and how.

In the meantime you can still choose to write a function that talks to the API server - nothing blocks it technically. You'll just need to be aware that your function isn't spec compliant, which could affect Crossplane's ability to successfully use it in future.

Alternatively, we could add wording today to allow talking to the API server only to get credentials. The risk there is that if we're not specific on _how_ to do that it runs the risk of folks inventing patterns that we might not want to support long term.

Reply:
> felt less like a specification concern and more like a documentation concern

I see value here in codifying the interaction points for the different operations and then using the documentation to detail the "how". This would also align with the previous statement on external systems which clearly calls our using MRs to mutate external systems.

You are correct in that the main concern is about IRSA capabilities for functions. The way I see this is the lack of this capability is a blocker towards more generalized adoption of functions which offer cloud specific capabilities. I also see it as something of a hard sell to tell engineers that within a single composition they would be providing multiple credential sources for the same cloud provider which is why I went for the ProviderConfig option for my functions but this is maybe a conversation for a different ticket.

I would like to avoid the functions I am writing diverging too far from the published spec for similar reasons and it would potentially be difficult to align them in the future as additional capabilities are opened up. Alignment is something I can partially solve on my own function interfaces though.

Regarding the potential addition of wording to cover credential retrieval, I kind of think this is opening a can of worms towards functions requesting broader api access unless there was a clearly defined path to retrieve them potentially backed by the function SDK - although inversely this risks pushing crossplane into supporting a pattern it does not want or need long term.